aspnetcore/fundamentals/http-requests/samples/Startup.cs

```diff
@@ -0,0 +1,149 @@
+        {
+            // basic usage
+            #region snippet1
+            services.AddHttpClient();
+            #endregion
+
+            // named client
+            #region snippet2
+            services.AddHttpClient("github", c =>
+            {
+                c.BaseAddress = new Uri("https://api.github.com/");
+                c.DefaultRequestHeaders.Add("Accept", "application/vnd.github.v3+json"); // Github API versioning
+                c.DefaultRequestHeaders.Add("User-Agent", "HttpClientFactory-Sample"); // Github requires a user-agent
+            });
+            #endregion
+
+            // typed client where configuration occurs in ctor
+            #region snippet3
+            services.AddHttpClient<GitHubService>();
+            #endregion
+
+            // typed client where configuration occurs during registration
+            #region snippet4
+            services.AddHttpClient<RepoService>(c =>
+            {
+                c.BaseAddress = new Uri("https://api.github.com/");
+                c.DefaultRequestHeaders.Add("Accept", "application/vnd.github.v3+json");
+                c.DefaultRequestHeaders.Add("User-Agent", "HttpClientFactory-Sample");
+            });
+            #endregion
+
+            #region snippet5
+            services.AddTransient<ValidateHeaderHandler>();
+
+            services.AddHttpClient("externalservice", c =>
+            {
+                c.BaseAddress = new Uri("https://localhost:5000/"); // assume this is an "external" service which requires an API KEY
+            })
+            .AddHttpMessageHandler<ValidateHeaderHandler>();
+            #endregion
+
+            #region snippet6
+            services.AddTransient<SecureRequestHandler>();
+            services.AddTransient<RequestDataHandler>();
+
+            services.AddHttpClient("clientwithhandlers")
+                // This handler is on the outside and called first during the request, last during the response.
+                .AddHttpMessageHandler<SecureRequestHandler>()
+                // This handler is on the inside, closest to the request being sent.
+                .AddHttpMessageHandler<RequestDataHandler>();
+            #endregion
+
+            #region snippet7
+            services.AddHttpClient<UnreliableEndpointCallerService>()
+                .AddTransientHttpErrorPolicy(p => p.WaitAndRetryAsync(3, _ => TimeSpan.FromMilliseconds(600)));
+            #endregion
+
+            #region snippet8
+            var timeout = Policy.TimeoutAsync<HttpResponseMessage>(TimeSpan.FromSeconds(10));
+            var longTimeout = Policy.TimeoutAsync<HttpResponseMessage>(TimeSpan.FromSeconds(30));
+
+            services.AddHttpClient("conditionalpolicy")
+            // Run some code to select a policy based on the request
+                .AddPolicyHandler(request => request.Method == HttpMethod.Get ? timeout : longTimeout);
+            #endregion
```

Comment:
It's worth calling out somewhere in the Polly part of the docs that users should treat polly policies as long-lived objects. Some policies only function correctly when they are shared or reused. We've worked to make this the idiomatic thing, but it should be covered in the docs as well.

Reply:
Yes, in general we do recommend defining the policy early on and reusing them. Policies are thread-safe and each call to `.Execute` or `.ExecuteAsync` establishes a new, independent underlying policy state. Certain policies, as you say, must be shared for maximum effectiveness, like the `CircuitBreaker`. I can update the Polly and HttpClientFactory doc to reiterate this fact and point to more detailed documentation on the subject.

Reply:
Actually, I believe this section of the document explains it clearly enough, do you agree? [Use case: scoping CircuitBreakers and Bulkheads](https://github.com/App-vNext/Polly/wiki/Polly-and-HttpClientFactory#use-case-scoping-circuitbreakers-and-bulkheads)